images/virtualization-artifact/pkg/controller/cvi/internal/source/object_ref_vd.go

```diff
@@ -0,0 +1,247 @@
+				return false, nil
+			default:
+				return false, err
+			}
+		}
+
+		err = ds.importerService.Protect(ctx, pod)
+		if err != nil {
+			return false, err
+		}
+
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = vicondition.Provisioning
+		condition.Message = "Import is in the process of provisioning to DVCR."
+
+		cvi.Status.Phase = virtv2.ImageProvisioning
+		cvi.Status.Progress = ds.statService.GetProgress(cvi.GetUID(), pod, cvi.Status.Progress)
+		cvi.Status.Target.RegistryURL = ds.statService.GetDVCRImageName(pod)
+
+		log.Info("Provisioning...", "progress", cvi.Status.Progress, "pod.phase", pod.Status.Phase)
+	}
+
+	return true, nil
+}
+
+func (ds ObjectRefVirtualDisk) CleanUpSupplements(ctx context.Context, cvi *virtv2.ClusterVirtualImage) (bool, error) {
+	supgen := supplements.NewGenerator(cc.CVIShortName, cvi.Name, ds.controllerNamespace, cvi.UID)
+
+	importerRequeue, err := ds.importerService.CleanUpSupplements(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	diskRequeue, err := ds.diskService.CleanUpSupplements(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	return importerRequeue || diskRequeue, nil
+}
+
+func (ds ObjectRefVirtualDisk) CleanUp(ctx context.Context, cvi *virtv2.ClusterVirtualImage) (bool, error) {
+	supgen := supplements.NewGenerator(cc.CVIShortName, cvi.Name, ds.controllerNamespace, cvi.UID)
+
+	importerRequeue, err := ds.importerService.CleanUp(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	diskRequeue, err := ds.diskService.CleanUp(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	return importerRequeue || diskRequeue, nil
+}
+
+func (ds ObjectRefVirtualDisk) getEnvSettings(cvi *virtv2.ClusterVirtualImage, sup *supplements.Generator) *importer.Settings {
+	var settings importer.Settings
+	importer.ApplyBlockDeviceSourceSettings(&settings)
+	importer.ApplyDVCRDestinationSettings(
+		&settings,
+		ds.dvcrSettings,
+		sup,
+		ds.dvcrSettings.RegistryImageForCVI(cvi),
+	)
+
+	return &settings
+}
+
+func (ds ObjectRefVirtualDisk) Validate(ctx context.Context, cvi *virtv2.ClusterVirtualImage) error {
+	if cvi.Spec.DataSource.ObjectRef == nil || cvi.Spec.DataSource.ObjectRef.Kind != virtv2.ClusterVirtualImageObjectRefKindVirtualDisk {
+		return fmt.Errorf("not a %s data source", virtv2.ClusterVirtualImageObjectRefKindVirtualDisk)
+	}
+
+	vd, err := ds.diskService.GetVirtualDisk(ctx, cvi.Spec.DataSource.ObjectRef.Name, cvi.Spec.DataSource.ObjectRef.Namespace)
+	if err != nil {
+		return err
+	}
+
+	if vd == nil || vd.Status.Phase != virtv2.DiskReady {
+		return NewVirtualDiskNotReadyError(cvi.Spec.DataSource.ObjectRef.Name)
+	}
+
+	if len(vd.Status.AttachedToVirtualMachines) != 0 {
+		vmName := vd.Status.AttachedToVirtualMachines[0]
+		vm, err := ds.diskService.GetVirtualMachine(ctx, vmName.Name, vd.Namespace)
+		if err != nil {
+			return err
+		}
+
+		if vm.Status.Phase == virtv2.MachineRunning {
```

Comment:
What if the VM is starting and will be running in one second during the copying from PVC to DVCR?

Reply:
Fixed.